### ISSUE TITLE
Testing best model checkpoint after training model with multiple GPUs

### ISSUE THREAD

**Issue by amorehead** [ashleve/lightning-hydra-template#445]
Hello.

First off, I am a huge fan of this project and want to thank its maintainers for their diligent efforts!

I noticed that in the default `train.py` example, the maintainers of this project would have users call `trainer.test()` on the best model checkpoint resulting from (single/multi) GPU training beforehand (if the `test` CLI argument is set to `True`). In such a case, I am a bit concerned about how Lightning will attempt to report its test results to users through the use of a logging tool like WandB. For example, I believe it is typically recommended to run a trained PyTorch/Lightning model on the test dataset(s) using only a single GPU to avoid any misreporting of test results.

Is such an option available with this project layout?

**Comment by ashleve** [comment]
Generally, metric values should be the same or only slightly different even when testing with multiple GPUs, since template uses torchmetrics which synchronize across all processes on each GPU.

**Comment by ashleve** [comment]
Generally, metric values should be the same or only slightly different even when testing with multiple GPUs, since template uses torchmetrics which synchronize across ddp processes.

https://torchmetrics.readthedocs.io/en/latest/pages/overview.html?highlight=ddp#metrics-in-distributed-data-parallel-ddp-mode

If you want to test on 1 GPU either way, you could just overwrite the `cfg.trainer.devices=1` in `train.py` after training and instantiate the Trainer again.

**Comment by amorehead** [comment]
@ashleve, thank you for your quick reply. With a solution that you suggested such as the following, should one expect their existing logging experiment (e.g., a run in WandB) to be preserved after creating a new `Trainer` instance this way? For example, my ultimate goal is to be able to train a model using multiple GPUs and to test on a single GPU, using a single execution of the same Python script, and to have all TorchMetrics logged to the same run in my WandB dashboard.

```
if cfg.get("test"):
        log.info("Starting testing!")
        ckpt_path = trainer.checkpoint_callback.best_model_path
        if ckpt_path == "":
            log.warning("Best ckpt not found! Using current weights for testing...")
            ckpt_path = None
        log.info(f"Best ckpt path: {ckpt_path}")
        # ensure testing only takes place on a single device
        cfg.trainer.devices = 1
        trainer = hydra.utils.instantiate(
            cfg.trainer, callbacks=callbacks, logger=logger
        )
        trainer.test(model=model, datamodule=datamodule, ckpt_path=ckpt_path)
```

**Comment by ashleve** [comment]
Yes. Information about the wandb experiment is stored in wandblogger object which is instantiated separately, so you just need to pass it to the trainer again:
```python
trainer = hydra.utils.instantiate(
            cfg.trainer, callbacks=callbacks, logger=logger
        )
```
(here `logger` is list of instantiated loggers)

**Comment by amorehead** [comment]
This is great to know. I will be looking to resume training runs (including their associated WandB runs) using this project template soon.

On a slightly different note, I tried to run my approach above for testing a trained model with a single GPU. It seems like there are no errors up to the last line, `trainer.test(model=model, datamodule=datamodule, ckpt_path=ckpt_path)`. At this point, the training script completely freezes and never terminates. Would you have any ideas where I should look for a solution to this problem? For the multi-GPU training part of this script, I am using `strategy=ddp, accelerator=gpu, devices=6, num_workers=4` for my `cfg.trainer` config.

**Comment by ashleve** [comment]
I'm not familiar with the problem, but on your place I would try to repeat this without using hydra (to confirm it's not a hydra-related issue), and then make a dedicated issue on lightning repository

**Comment by amorehead** [comment]
Thanks for your feedback, @ashleve!
